### DISCUSSION TITLE
Is there a way to get effective user permissions on a list?

### DISCUSSION THREAD

**Discussion by SuleymanA** [pnp/pnpcore#910] [Q&A] [answered]
I am struggling a lot to find a solution for getting the effective user permissions on a list. I want to check on a list whether the user as permission to create an item. On SPFX side this is really easy to do by calling "getusereffectivepermissions" on the list. I didn't see any build in option in PnP core List object so I tried doing the rest call but I am not able to parse it a way that I can check the permissions. What I have now is this but the parsing seems to be going wrong:

            var apiRequest = new ApiRequest(ApiRequestType.SPORest, "/_api/web/lists/getbytitle('" + listTitle + "')/getusereffectivepermissions(@u)?@u='" + Uri.EscapeDataString("i:0#.f|membership|" + userUPN) + "'");
            var response = pnpContext.Web.ExecuteRequest(apiRequest);

            var userPermissions = JsonConvert.DeserializeObject<Microsoft.SharePoint.Client.BasePermissions>(response.Response);
            return userPermissions.Has(Microsoft.SharePoint.Client.PermissionKind.AddListItems);

This is done in an Azure Function. Any help is much appriciated!!

**Comment by jansenbe** [comment, marked as answer]
@SuleymanA : you need to manually process the json, the internal JSON processing for REST/Graph is not exposed to the outside. We can add this a feature in the library, can you make an issue describing the ask so that we can track it for future inclusion. 

Thanks for using PnP Core SDK!

**Reply by SuleymanA** [reply]
It would be great to have a decidated method in the list object to check current user permissions. I managed to do it by what seems a dirty workaround. I couldn't find a way to use IBasepermissions in a proper way (still looking for a better way to do it): 

            var apiRequest = new ApiRequest(ApiRequestType.SPORest, "/_api/web/lists/getbytitle('" + listTitle + "')/getusereffectivepermissions(@u)?@u='" + Uri.EscapeDataString("i:0#.f|membership|" + userUPN) + "'");
            var response = pnpContext.Web.ExecuteRequest(apiRequest);

            var jo = JObject.Parse(response.Response);
            var data = (JObject)jo["d"]["GetUserEffectivePermissions"];
          
            IBasePermissions test = pnpContext.Web.EffectiveBasePermissions;

            test.High = long.Parse(data["High"].ToString());
            test.Low = long.Parse(data["Low"].ToString());

            return test.Has(PermissionKind.AddListItems);

Using PnP Core has been a bit of a struggle. A lot of things seem to be done much more complicated than it was before. Simple things like reading and writing list data. Comparing to PnP Framework and especially to SPFX development where it flows so much smoother.

**Reply by jansenbe** [reply]
@SuleymanA : I added issue https://github.com/pnp/pnpcore/issues/912 to track your ask for including getusereffectivepermissions. 

Would like to understand what parts you do find much more complicated, here's our docs for reading and adding list items: https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html

**Reply by SuleymanA** [reply]
Thx for creating feature request.

I am building the front end of the application in SPFX using mainly SP REST API which feels so much easier to use with stuff like expanding lookup or people fields, not using CAML for querying and much easier to update listItems (like a (multi-)people field. In PnP I need to get the field definition (didn't find a way to instantiate an empty NewFieldValueCollection) before I can assign values to it).

It just feels like a lot of code that is not helping me and makes readability bad because of the bulky way of querying and field handling. With rest api I just need a line of code to query, in PNP Core its a whole lot more because of use of CAML and the usage of this thing "LoadListDataAsStreamAsync" for querying.

On documentation, I think it doenst have simple enough examples to follow. For example on field usage: In one code section the example contains pieces for creating a field, adding value and updating value. Would be a lot easier to have seperate examples (one example for creating fields with all types, one for updating value for all types, etc. this fits reallife scenarions much better). For me it was not clear that I needed to get the field first and then on the field create a new NewFieldValueCollection before I can set the values for multi-people field.

Sorry if I sound too negative, I am just having a hard time adjusting, probably also just lack of experience with PNP Core.

**Reply by jansenbe** [reply]
@SuleymanA : issue #912 has been resolved, please let us know how it works for you.

Regarding working with list items and fields: we did simplify our model a bit, making it easier for working with FieldValue and FieldValueCollection objects: https://github.com/pnp/pnpcore/issues/915
